Libraries/LLM/Evaluate.swift

```diff
@@ -78,54 +34,186 @@ public struct GenerateParameters: Sendable {
     }
 }
 
+struct SampleContext {
+
+    let temp: MLXArray
+    let topP: MLXArray
+    let useTopP: Bool
+    let useArgMax: Bool
+
+    init(parameters: GenerateParameters) {
+        self.temp = MLXArray(parameters.temperature)
+        self.topP = MLXArray(parameters.topP)
+        self.useTopP = parameters.topP > 0 && parameters.topP < 1
+        self.useArgMax = parameters.temperature == 0
+    }
+
+    private let compiledTopPSampling: (MLXArray, MLXArray, MLXArray) -> MLXArray = {
+        compile(inputs: [MLXRandom.globalState], outputs: [MLXRandom.globalState]) {
+            logits, topP, temp in
+            let probs = softmax(logits / temp, axis: -1)
+            let sortedIndices = argSort(probs, axis: -1)
+
+            // probs shape is [B,V] and after take it will be [1, B, V], so we squeeze it back to [B, V]
+            let sortedProbs = take(probs, sortedIndices, axis: -1).squeezed(axis: 0)
+
+            let cumulativeProbs = cumsum(sortedProbs, axis: -1)
+
+            let topProbs = MLX.where(
+                cumulativeProbs .> (1 - topP), sortedProbs, zeros(like: sortedProbs))
+
+            let sortedToken = categorical(log(topProbs))
+            return sortedIndices.squeezed(axis: 0)[sortedToken]
+        }
+    }()
+
+    private let compiledCategorical: (MLXArray, MLXArray) -> MLXArray = {
+        compile(inputs: [MLXRandom.globalState], outputs: [MLXRandom.globalState]) { logits, temp in
+            categorical(logits * (1 / temp))
+        }
+    }()
+
+    private func topPSampling(logits: MLXArray) -> MLXArray {
+        var logits = logits
+        if logits.dtype == .bfloat16 {
+            logits = logits.asType(.float32)
+        }
+
+        return compiledTopPSampling(logits, topP, temp)
+    }
+
+    func sample(logits: MLXArray) -> MLXArray {
+        if useArgMax {
+            return argMax(logits, axis: -1)
+        } else {
+            if useTopP {
+                return topPSampling(logits: logits)
+            } else {
+                return compiledCategorical(logits, temp)
+            }
+        }
+    }
+}
+
+/// Encapsulaton of the repetitionPenalty
+struct RepetitionContext: Sendable {
```

Comment:
Refactored out the repetition context code for encapsulation -- I found that this was missing the token append line.